js/src/logger.ts

```diff
@@ -745,42 +748,80 @@ export class Attachment {
   private readonly dataDebugString: string;
 
   /**
    * Construct an attachment.
    *
-   * @param data A string representing the path of the file on disk, or a
+   * @param param A parameter object with:
+   *
+   * `data`: A string representing the path of the file on disk, or a
    * `Blob`/`ArrayBuffer` with the file's contents. The caller is responsible
    * for ensuring the file/blob/buffer is not modified until upload is complete.
    *
-   * @param filename The desired name of the file in Braintrust after uploading.
+   * `filename`: The desired name of the file in Braintrust after uploading.
    * This parameter is for visualization purposes only and has no effect on
    * attachment storage.
    *
-   * @param contentType The MIME type of the file.
+   * `contentType`: The MIME type of the file.
    *
-   * @param state (Optional) For internal use.
+   * `key`: (Optional) For internal use. Non-empty key to the attachment in
+   * blob storage, if this object represents an existing attachment. It is
+   * recommended to use the convenience method {@link Attachment.fromReference}
+   * instead of setting this parameter directly.
+   *
+   * `state`: (Optional) For internal use.
    */
-  constructor({ data, filename, contentType, state }: AttachmentParams) {
+  constructor({ data, filename, contentType, key, state }: AttachmentParams) {
     this.reference = {
       type: BRAINTRUST_ATTACHMENT,
       filename,
       content_type: contentType,
-      key: newId(),
+      key: key || newId(),
     };
     this.state = state;
     this.dataDebugString = typeof data === "string" ? data : "<in-memory data>";
 
-    this._data = this.initData(data);
-    this.uploader = this.initUploader();
+    this.isUpload = !key;
+    if (this.isUpload) {
+      this._data = this.initData(data);
+      this.uploader = this.initUploader();
+    } else {
+      this._data = this.initDownloader();
+      this.uploader = this.initStatusGetter();
+    }
+  }
+
+  /**
+   * Create an `Attachment` object from the given reference. Used for reading
+   * the contents of a previously uploaded attachment.
+   *
+   * @param reference The `AttachmentReference` that should be read by the
+   * `Attachment` object.
+   * @param state (Optional) For internal use.
+   * @returns The new `Attachment` object.
+   */
+  static fromReference(
+    reference: AttachmentReference,
+    state?: BraintrustState,
+  ): Attachment {
+    return new Attachment({
+      data: new Blob(), // Placeholder value will be ignored by the callee.
+      filename: reference.filename,
+      contentType: reference.content_type,
+      key: reference.key,
+      state,
+    });
   }
 
   /**
    * On first access, (1) reads the attachment from disk if needed, (2)
    * authenticates with the data plane to request a signed URL, (3) uploads to
-   * object store, and (4) updates the attachment.
+   * object store, and (4) updates the attachment. If this object represents an
```

Comment:
I wonder if it'd be a better DX if we more clearly separate the methods for upload-mode and download-mode. Or maybe even separate the classes since there are so few public methods to begin with.

I just feel like it'd look a little weird to see code like

```
const attachment = Attachment.fromReference(ref);
const status = attachment.upload();
```

and not think it's uploading something. Like what if we have a separate `ReadonlyAttachment` class that handles all of this new stuff (analogous to `ReadonlyExperiment`)